send_email: Add setting to control max connection lifetime.

shubham-padia opened the pull request:
Fixes #21900.

I have not tested the code with a real SMTP connection, should the testing code here be sufficient or is there a need of me to test it out in a real world scenario?

**Changes from the original plan**:

- The default for the setting is set to None, in order to avoid disruption to installs used to the existing behaviour. 
- When the setting is set to 0, we do not actively close the connection when the email is sent, but we open a new connection everytime we need to send a new one.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

shubham-padia commented on zproject/email_backends.py:
```diff
@@ -108,3 +109,9 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
+class CustomSMTPEmailBackend(EmailBackend):
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__(**kwargs)  # nocoverage
+        self.opened_at = timezone_now()  # nocoverage
```
Should it be okay to skip coverage in cases where the class is being mocked by MagicMock?

timabbott replied:
I think you can use the `time_machine` library to mock time without mocking this function, and that'll remove the ened to use `nocoverage`.

shubham-padia replied:
Since the mocked object was needed for the original test, I've added the following extra lines to the test file to add coverage:


```js
        # Test opened_at is being set to timezone_now
        with time_machine.travel(
            timezone_now(),
            tick=False,
        ):
            custom_backend = CustomSMTPEmailBackend()
            self.assertEqual(custom_backend.opened_at, timezone_now())
 ```

alexmv commented:
> The default for the setting is set to None, in order to avoid disruption to installs used to the existing behaviour.

Essentially all install that are not Zulip Cloud do not need this performance optimization, and we certainly do get big reports about how the current behaviour breaks things for users -- please do change this PR to make the default be non-persistent connections.  Backwards compatibility is not a concern here.

alexmv commented on zproject/email_backends.py:
```diff
@@ -108,3 +109,9 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
         return num_sent
+
+
+class CustomSMTPEmailBackend(EmailBackend):
```
Yes, it's custom -- but this name doesn't tell me _what about it is custom_.  I think we can come up with a better name that describes what this does. :)

alexmv replied:
And if we're going to use this as our EmailBackend -- can we move the connection-tracking logic from `initialize_connection` onto this class?  Having a `isinstance(connection, CustomSMTPEmailBackend)` where _we wrote the class_ is a smell that we're possibly not actually using good object orientation.

shubham-padia replied:
This will our new `SMTP` email backend but not BaseEmailBackend. I'm wondering whether it would be appropriate to handle `EmailLogBackend` logic in a SMTP email backend class.

alexmv replied:
I think you accidentally a word?

`EmailLogBackend` logic should remain separate.

shubham-padia replied:
I've moved the logic to `PersistentSMTPEmailBackend` in a seperate commit, so it's easier to review and make changes for me if required, we can squash it before merging if we desire to have them in the same commit

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -104,6 +113,64 @@ def test_initialize_connection(self) -> None:
+            initialize_connection(backend)
+
+        self.assertEqual(backend.close.call_count, 1)
+        self.assertEqual(backend.open.call_count, 3)
```
...however, we do call `open` because that's how we check if the connection is _already_ open.  This probably deserves a comment because you just above say "we will not open a new connection" but then assert that we ... called `open`.  Which appears to contradict the above comment, and thus deserves its own comment.

shubham-padia replied:
Added the comment and fixed!

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -104,6 +113,64 @@ def test_initialize_connection(self) -> None:
+        self.assertEqual(backend.open.call_count, 3)
+        self.assertEqual(backend.connection.noop.call_count, 1)
+
+        # Old connection is open, we will not open a new connection because enough time has not elapsed.
```
I think it's slightly more idiomatic, and readable, to say:
```suggestion
        # Old connection is open, we will not open a new connection because not enough time has elapsed.
```

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -104,6 +113,64 @@ def test_initialize_connection(self) -> None:
+        backend.open.return_value = False
+        backend.connection.noop.return_value = [250]
+
+        # Old connection is open, but we still reopen a new connection because max connection lifetime is 0.
+        with mock.patch.object(backend, "opened_at", timezone_now(), create=True), self.settings(
+            EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=0
+        ):
+            initialize_connection(backend)
```
All of these are _probably_ safe -- but we're not using `time_machine` here, so we don't actually know what time it will be when `initialize_connection` checks.  Again, in this case it shouldn't matter what time it is, but it's good practice to _always_ use `time_machine` or similar when we're testing things that are time-dependent.

shubham-padia replied:
Fixed

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -104,6 +113,64 @@ def test_initialize_connection(self) -> None:
+        self.assertEqual(backend.connection.noop.call_count, 1)
+
+        # Old connection is open, we will not open a new connection because enough time has not elapsed.
+        with mock.patch.object(backend, "opened_at", timezone_now(), create=True), self.settings(
+            EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=5
+        ):
+            initialize_connection(backend)
```
This needs a `time_machine`, since there _is no guarantee_ that the `initialize_connection` is called within 5 minutes of when you set that mock.  It would be totally unreasonable for the computer to wander off and make tea between statements, but stranger things have happened.  Use `time_machine`.

shubham-padia replied:
Fixed

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -104,6 +113,64 @@ def test_initialize_connection(self) -> None:
+        self.assertEqual(backend.open.call_count, 6)
+        self.assertEqual(backend.connection.noop.call_count, 2)
+
+        # Test opened_at is being set to timezone_now
+        with time_machine.travel(
+            timezone_now(),
+            tick=False,
+        ):
+            custom_backend = CustomSMTPEmailBackend()
+            self.assertEqual(custom_backend.opened_at, timezone_now())
+
```
It's far more interesting to put this check inside each of your above blocks -- most notably the previous test, where you expect it to re-open the connection, and we _need_ that to reset the `opened_at`.  And using `time_machine` on all of those `initialize_connection` calls will let you know what time to expect it to be.

alexmv reviewed (commented):
> Added # nocoverage comments to CustomSMTPEmailBackend.__init__ since the class is being mocked by MagicMock.

This comment in the commit message is no longer accurate.

alexmv commented:
It looks like you addressed some of the review comments (the detault) but not others (using `time_machine` in tests, moving logic into the `TimeAwareSMTPEmailBackend` class instead of using `isinstance`).  Please ping me when this is ready for another review.

Two quick notes -- first, remember to update your commit message for the new defaults, etc.  Second, `PersistentSMTPEmailBackend` seems like a more precise name -- while it is true that it is time aware, that doesn't help me understand the greater purpose of the class.

shubham-padia commented:
@alexmv the previous push wasn't ready for review, I've addressed the requested changes and it's ready for review now.

Modifying the tests also led me to discover that my initial logic of setting `opened_at` at init was wrong, since the connection is not opened on init for `SMTPBackend`, I have moved that logic to `open` now.

shubham-padia commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
+                # backend.connection was overwritten in a previous
+                # `initialize_connection` call in this test.
+                backend.connection = mock.MagicMock(spec=SMTP)
+                backend.connection.noop.return_value = [250]
```
I wasn't able to mock the chained call using `with mock.patch.object()` resulting in this bit of ugly-ish code, let me know if there's a better way to do this, thanks!

shubham-padia commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
-        initialize_connection(backend)
-        # 2 more calls to open, 1 more call to noop and 1 call to close
-        self.assertEqual(backend.open.call_count, 3)
-        self.assertEqual(backend.connection.noop.call_count, 2)
```
I had to remove this check, since backend.connection was getting overwritten in the previous `initialize_connection` call and it was a regular method in this line instead of a mocked method

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
+        ) as _open_call, mock.patch.object(backend, "close", wraps=backend.close) as close_call:
+            # Old connection is open, but we still reopen a new connection because max connection lifetime is 0.
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
             initialize_connection(backend)
-        # 3 more calls to open as we try 3 times before giving up
-        self.assertEqual(backend.open.call_count, 6)
+
+            self.assertEqual(close_call.call_count, 1)
+            self.assertEqual(_open_call.call_count, 2)
+            self.assertEqual(backend.opened_at, timezone_now())
```
This is mentioned below -- but this is an unusual case which isn't testing what you think.  Specifically, this is the case where despite the max lifetime = 0, but the connection was left open from the last call, which is not generally what we expect. It does call `open` (to check for being open), `close`, and `open` -- but it would also do that same sequence if the connection _started off closed_, which is not correct.  You should add a test for the "happy path" where `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTE = 0` but the connection starts off closed.  It should call one `open` and one `close`.

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
+                initialize_connection(backend)
+
+            self.assertEqual(close_call.call_count, 1)
+            # The increase in open.call_count by 1 is caused by us calling connection.open()
+            # to check whether an open connection already exists. In case of actually opening
+            # a new connection, the call count will increase by 2.
```
It might be nicer to (thoughout, not just here) use `reset_mock()` so we can be explicit about the number of calls _here_, rather than having to diff to the previous part of the test.

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
+            # a new connection, the call count will increase by 2.
+            self.assertEqual(_open_call.call_count, 3)
+
+            # Old connection is open, we will not open a new connection because not enough time has elapsed.
```
Well, 0 time has passed, which might theoretically be an edge case.  I'd make this travel 3 minutes, to make it clearly less than 5 but more than 0.

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
+            self.assertEqual(_open_call.call_count, 6)
+
+    @time_machine.travel(timezone_now(), tick=False)
+    def test_max_connection_lifetime(self) -> None:
```
Maybe break this up into a test for `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTE=undef`, one for `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTE=0`, and one for `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTE=5` ?

alexmv commented on zproject/email_backends.py:
```diff
@@ -108,3 +110,46 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
+
+    @override
+    def open(self, **kwargs: Any) -> bool | None:
+        is_opened = self._open()
+        status = None
+
+        time_elapsed = (timezone_now() - self.opened_at).seconds / 60
+        if (
+            settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES is None
+            or time_elapsed <= settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES
+        ):
+            # No-op to ensure that we don't return a connection that has been
+            # closed by the mail server.
+            try:
+                assert self.connection is not None
+                status = self.connection.noop()[0]
+            except Exception:
+                pass
+
+        if status is None or status != 250:
+            # Close and connect again.
+            self.close()
+            is_opened = self._open()
```
This looks like if we have `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES = 0`:

1. We open the connection on line 126.
2. We calculate that it's been 0.00001 seconds (or whatever) since we opened it, so we don't drop into the `if`
3. `status is None` so we close the connection and re-open it

Which means we're wasting our time and the SMTP server's connecting _twice_ for every message we send.

shubham-padia replied:
Fixed, this code block does not run on that setting being set to 0.

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,185 @@ def test_limited_to_length(self) -> None:
+            close_call.reset_mock()
+            _open_call.reset_mock()
+
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
```
We shouldn't be doing any noops here?  If you're going to set up this mock, check that it _isn't_ called

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,185 @@ def test_limited_to_length(self) -> None:
+    @override_settings(EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=0)
+    def test_max_connection_lifetime_zero(self) -> None:
+        backend = PersistentSMTPEmailBackend()
+        backend.open()
```
We should do this step inside of where we set up our mocks.  We can thus test how things work from an initial state -- which should be 0 noops, 0 closes, and 1 open.

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,185 @@ def test_limited_to_length(self) -> None:
+            close_call.reset_mock()
+            # Old connection is closed, but we still reopen a new connection because max connection lifetime is 0.
+            initialize_connection(backend)
+            self.assertEqual(close_call.call_count, 1)
+            self.assertEqual(_open_call.call_count, 2)
```
This should be 0 closes, 0 noops, and 1 open.

shubham-padia replied:
Noops are not the easiest to count when reopening a connection, since backend.connection gets replaced when initializing connection. I haven't added a call count for noop in this case.

alexmv commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,185 @@ def test_limited_to_length(self) -> None:
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
+
+            # Old connection is open, we will not open a new connection because max connection lifetime is None.
```
...but also no time has passed, so this test isn't really testing how the lifetime being None interacts with the march of time?

shubham-padia replied:
Fixed!

shubham-padia commented on zerver/tests/test_send_email.py:
```diff
@@ -70,39 +74,185 @@ def test_limited_to_length(self) -> None:
+        ):
+            # Old connection is open, but we still reopen a new connection because max connection lifetime is 0.
+            initialize_connection(backend)
+            self.assertEqual(close_call.call_count, 1)
+            self.assertEqual(_open_call.call_count, 2)
```
I don't get this combination, 1 close and 0 noops is fine, but in case of an existing connection being open, first `_open` call will be used to check the existence of the connection, then we'll close the open connection and reopen the connection using `_open` leading to two _open calls

shubham-padia replied:
Same reason for noop count as the other comment. For _open, it will be called once to check whether the connection exists or not at the beginning of the `open` function and for a second time after calling `.close()`.

alexmv commented on zproject/email_backends.py:
```diff
@@ -108,3 +110,46 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
+            except Exception:
+                pass
+
+        # In case of EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES = 0, we
+        # will always return a newly opened connection. The default
+        # smtp.py implementation of send_messages closes the connection
+        # after sending the emails if the connection was a new
+        # connection i.e. is_opened is True.
```
Doesn't this imply that we can never have a persistent connection?

That is, even if `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES is None`, the _first_ time we run this, we're opening a new connection and thus returning `True`.  Which by this comment would imply that it will also get closed immediately after sending one message?

This comment, or a clarified version of it if accurate, is probably better at the top, inside the first `if is_opened`.

alexmv replied:
OK, the solution to my mystery of two comments above is that:
1. If we're in the `email_senders` worker, we call `initialize_connection` which calls `open()`.
2. The default `send_messages` implementation calls `open()` again -- hence necessitating the check for `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES != 0` above, since the `initialize_connection` means that even non-persistent connections will get `open()` called on them multiple times, which is quite non-intuitive.
3. So the above noop block is skipped for non-persistent connections, leaving `status = None`.
4. ...sooooo we hit the block below this and close the connection and re-open it, uselessly.
5. ...but that bug _also_ means that we return `True` from the second `open()` call, which means that `send_messages()` will then call `close()` for us.  Which is correct behaviour (albeit with the extra open/close)
6. ...but that _also also_ means that if we have a persistent connection, when we time out and hit the block below this, we'll return `True`, causing `send_messages()` to think it needs to clean up.  Which means it'll send one message over the newly-opened connection, and then close it.

So this logic is unfortunately still not correct.  I am beginning to suspect that we may want to _only use this subclass_ if persistent connections are used (`EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES != 0`) because it's not necessary in the case where it's `== 0`, and it complicates both the logic, and how easy the code is to reason about.

alexmv commented on zproject/email_backends.py:
```diff
@@ -108,3 +110,46 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
+
+        status = None
+        time_elapsed = (timezone_now() - self.opened_at).seconds / 60
+        if settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES != 0 and (
```
Through your current logic of returning early if `is_opened` is True, the `!= 0` check here is purely defensive, right?

This has been enough back-and-forth over this logic that it's very definitely going to need a comment in it working through why the logic is correct. :)

alexmv commented on zproject/email_backends.py:
```diff
@@ -108,3 +110,46 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
+            self.close()
+            is_opened = self._open()
+
+        return is_opened
```
Per the above, I suspect this always needs to be False here -- or we need to be overriding the behaviour of `send_messages()` or `close()` so that a reconnection doesn't trigger the "oh, OK, I should close this, too" behaviour from the superclass.
